components/authz-service/storage/v2/postgres/postgres.go

```diff
@@ -1078,21 +1079,74 @@ func (p *pg) DeleteRule(ctx context.Context, id string) error {
 		return p.processError(err)
 	}
 
-	res, err := tx.ExecContext(ctx,
-		`DELETE FROM iam_project_rules WHERE id=$1 AND projects_match_for_rule(project_id, $2);`,
-		id, pq.Array(projectsFilter),
-	)
+	assocMap, err := p.getMapOfRuleAssociations(ctx, tx, id)
 	if err != nil {
 		return p.processError(err)
 	}
 
-	count, err := res.RowsAffected()
-	if err != nil {
-		return p.processError(err)
-	} else if count == 0 {
+	_, ruleStaged := assocMap[pgStaged]
+	_, ruleApplied := assocMap[pgApplied]
+
+	if !ruleStaged && !ruleApplied {
 		return storage_errors.ErrNotFound
 	}
 
+	if ruleApplied {
+		if ruleStaged {
+			res, err := tx.ExecContext(ctx,
+				`UPDATE iam_staged_project_rules
+					SET deleted=true
+					WHERE id=$1 AND projects_match_for_rule(project_id, $2);`,
+				id, pq.Array(projectsFilter),
+			)
+			if err != nil {
+				return p.processError(err)
+			}
+			err = p.singleRowResultOrNotFoundErr(res)
+			if err != nil {
+				return err
+			}
+		} else {
+			// check if filtered by project filter
```

Comment:
The reason I called it out there is the only reason I have that extra `SELECT` query is to apply the project filter; whereas the other if branches can apply the project filter as part of the actual data update queries (since the `iam_staged_project_rules` entries already exist). I agree the comment isn't adding much though. Will remove.